### PR TITLE
sandbox-perf: Fix broken benchmarks and make sure they stay fixed.

### LINKED ISSUE #4229: Ignored failures in generating benchmark code in `//ledger/sandbox-perf`
There are failures in generating benchmark code in `//ledger/sandbox-perf`, which are ignored.

I don't know the impact of this, but I imagine it's not great.

```
INFO: From Generating benchmark code for //ledger/sandbox-perf:sandbox-perf_codegen:
JMH benchmark generation: JMH Benchmark generator failed
JMH benchmark generation: Benchmark classes should not be final.
   [com.digitalasset.platform.sandbox.perf.LargeTransactionBench]
JMH benchmark generation: The instantiated @State class cannot be abstract.
   [com.digitalasset.platform.sandbox.perf.PerfBenchState]
```

### PR DESCRIPTION
**Pull request by SamirTalwar**
The `sandbox-perf` build has been failing for a while with the following
errors:

```
INFO: From Generating benchmark code for //ledger/sandbox-perf:sandbox-perf_codegen:
JMH benchmark generation: JMH Benchmark generator failed
JMH benchmark generation: Benchmark classes should not be final.
   [com.digitalasset.platform.sandbox.perf.LargeTransactionBench]
JMH benchmark generation: The instantiated @State class cannot be abstract.
   [com.digitalasset.platform.sandbox.perf.PerfBenchState]
```

However, these errors are ignored; running the benchmarks runs the
`AcsBench` benchmark and ignores the fact `LargeTransactionBench` and
`SimpleBench` failed to compile.

I've fixed the errors by making sure that classes are not final, and
that `SimpleBench` uses a concrete state class.

This also introduces a cheap patch to the Scala JMH Bazel rules that
makes sure they fail if there are any errors.

I'm not really sure how to patch the Bazel rules properly, but someone
else might have an idea. :-)

Closes #4229.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: bazel_tools/scala-fail-jmh-build-on-error.patch [hunk 1]
```diff
@@ -0,0 +1,46 @@
+diff --git a/src/scala/io/bazel/rules_scala/jmh_support/BenchmarkGenerator.scala b/src/scala/io/bazel/rules_scala/jmh_support/BenchmarkGenerator.scala
```
Comment (aherrmann-da) [inline bazel_tools/scala-fail-jmh-build-on-error.patch]: That looks good to me, would be great to upstream to https://github.com/bazelbuild/rules_scala

Comment (SamirTalwar) [reply bazel_tools/scala-fail-jmh-build-on-error.patch]: Will do!

Comment (SamirTalwar) [reply bazel_tools/scala-fail-jmh-build-on-error.patch]: Opened: https://github.com/bazelbuild/rules_scala/pull/977

**Comment by SamirTalwar-DA** [comment]
FYI @aherrmann-da: this has now been patched upstream in https://github.com/bazelbuild/rules_scala/pull/977, so when we upgrade Bazel to v2.0, we can potentially upgrade `rules_scala` too and delete the patch.

**Comment by aherrmann-da** [comment]
@SamirTalwar-DA Awesome, thanks for upstreaming this!

**Pull request merged**